File: php-bpm-finder/src/BpmFinder.php
<?php

namespace BpmFinder;

class BpmFinder
{
    /**
     * Calculate BPM from array of tap timestamps in milliseconds.
     */
    public static function calculateBpmFromIntervals(array $timestampsMs): ?int
    {
        if (count($timestampsMs) < 2) {
            return null;
        }

        $intervals = [];
        for ($i = 1; $i < count($timestampsMs); $i++) {
            $intervals[] = $timestampsMs[$i] - $timestampsMs[$i - 1];
        }

        $avgInterval = array_sum($intervals) / count($intervals);
        if ($avgInterval <= 0) {
            return null;
        }

        $bpm = (int) round(60000 / $avgInterval);
        return ($bpm >= 30 && $bpm <= 300) ? $bpm : null;
    }

    /**
     * Calculate delay times in milliseconds for a given BPM.
     */
    public static function calculateDelayTimes(float $bpm): array
    {
        if ($bpm <= 0) {
            throw new \InvalidArgumentException('BPM must be greater than 0');
        }

        $quarterMs = 60000.0 / $bpm;

        return [
            'quarter_note_ms' => round($quarterMs, 2),
            'eighth_note_ms'  => round($quarterMs / 2.0, 2),
            'sixteenth_note_ms' => round($quarterMs / 4.0, 2),
            'dotted_eighth_ms' => round($quarterMs * 0.75, 2),
            'triplet_eighth_ms' => round(($quarterMs * 2.0) / 3.0, 2),
        ];
    }

    /**
     * Calculate recommended reverb pre-delay times in milliseconds.
     */
    public static function calculateReverbPredelay(float $bpm): array
    {
        $delayTimes = self::calculateDelayTimes($bpm);
        return [
            'tight_predelay_ms'   => round($delayTimes['sixteenth_note_ms'] / 2.0, 2),
            'natural_predelay_ms' => $delayTimes['sixteenth_note_ms'],
            'spacious_predelay_ms' => $delayTimes['eighth_note_ms'],
        ];
    }
}

File: dotnet-bpm-finder/BpmFinder.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace BpmFinder
{
    public class DelayTimes
    {
        public double QuarterNoteMs { get; set; }
        public double EighthNoteMs { get; set; }
        public double SixteenthNoteMs { get; set; }
        public double DottedEighthMs { get; set; }
        public double TripletEighthMs { get; set; }
    }

    public class ReverbPredelay
    {
        public double TightPredelayMs { get; set; }
        public double NaturalPredelayMs { get; set; }
        public double SpaciousPredelayMs { get; set; }
    }

    public static class BpmCalculator
    {
        public static int? CalculateBpmFromIntervals(IReadOnlyList<double> timestampsMs)
        {
            if (timestampsMs == null || timestampsMs.Count < 2) return null;

            var intervals = new List<double>();
            for (int i = 1; i < timestampsMs.Count; i++)
            {
                intervals.Add(timestampsMs[i] - timestampsMs[i - 1]);
            }

            double avgInterval = intervals.Average();
            if (avgInterval <= 0) return null;

            int bpm = (int)Math.Round(60000.0 / avgInterval);
            return (bpm >= 30 && bpm <= 300) ? (int?)bpm : null;
        }

        public static DelayTimes CalculateDelayTimes(double bpm)
        {
            if (bpm <= 0) throw new ArgumentOutOfRangeException(nameof(bpm), "BPM must be greater than 0");

            double quarterMs = 60000.0 / bpm;

            return new DelayTimes
            {
                QuarterNoteMs = Math.Round(quarterMs, 2),
                EighthNoteMs = Math.Round(quarterMs / 2.0, 2),
                SixteenthNoteMs = Math.Round(quarterMs / 4.0, 2),
                DottedEighthMs = Math.Round(quarterMs * 0.75, 2),
                TripletEighthMs = Math.Round((quarterMs * 2.0) / 3.0, 2)
            };
        }

        public static ReverbPredelay CalculateReverbPredelay(double bpm)
        {
            var delayTimes = CalculateDelayTimes(bpm);
            return new ReverbPredelay
            {
                TightPredelayMs = Math.Round(delayTimes.SixteenthNoteMs / 2.0, 2),
                NaturalPredelayMs = delayTimes.SixteenthNoteMs,
                SpaciousPredelayMs = delayTimes.EighthNoteMs
            };
        }
    }
}
